Install Docs Update

kenafoster opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://nebari.dev/community
-->

## Reference Issues or PRs

<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create a link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

## What does this implement/fix?

_Put a `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features not to work as expected)
- [ X ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

This PR intends to update the quick start and installation guides with the following clarifications:

- Clarify the distinction between installing the Nebari package and deploying a Nebari system (e.g. the Nebari package is the "Nebari deployer" and is used to stand up Nebari deployments on local or remote systems).   
- Clarify that the version of the "Nebari deployer" package (e.g. pip install nebari==1.2.3) must match the version of Nebari specified in nebari-config.yaml   
- Clarify that all Nebari extensions installed in the environment used to deploy Nebari will themselves be deployed along with Nebari, and that this is the mechanism used to specify and deploy Nebari extensions   
- Clarify that there is no mechanism to prevent deployment of locally-installed Nebari extensions, nor is there any mechanism to un-install an extension from a deployment.   
- Add "Best Practices" for creating a known Python environment used for deployment. Add example steps to use conda for this purpose
- Move the "Validate (optional)" and "Render (optional)" sections to before the "Deploy" section
- Summarize the purpose of the /stage/ directory, when it can be deleted, and why deleting it can resolve certain deployment errors.

## Testing

- [ ] Did you test the pull request locally?
- [ ] Did you add new tests?

## Documentation

<!--
  Where is this feature or API documented?

  - If docs exist:
    - Update any references, if relevant. This includes Guides and Nebari Internals docs.
  - If no docs exist:
    - Create a stub for documentation, including bullet points for how to use the feature, code snippets (including from
      happy path tests), etc.
We want to ensure that the content produced for Nebari is accessible; if you are making significant contributions,
please address the access-centred guidelines in your content and complete our checklist. Thanks to @isabela-pf for this checklist :)
-->

### Access-centered content checklist

#### Text styling

- [ ] The content is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [ ] If there are headers, they use the proper header tags (with only one level-one header: `H1` or `#` in markdown).
- [ ] All links describe where they link to (for example, check the [Nebari website](https://nebari.dev/)).
- [ ] This content adheres to the Nebari style guides.

#### Non-text content

- [ ] All content is represented as text (for example, images need alt text, and videos need captions or descriptive transcripts).
- [ ] If there are emojis, there are not more than three in a row.
- [ ] Don't use [flashing GIFs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [ ] If the content were to be read as plain text, it still makes sense, and no information is missing.

## Any other comments?

<!--
Please be aware that we are a loose team of volunteers, so patience is necessary;
assistance handling other issues is very welcome.
We value all user contributions. If we are slow to review, either the pull request needs some benchmarking, tinkering,
convincing, etc., or the reviewers are likely busy. In either case,
we ask for your understanding during the
review process.
Thanks for contributing to Nebari 🙏🏼!
-->

viniciusdc commented on docs/docs/explanations/advanced-configuration.md:
```diff
@@ -283,19 +283,32 @@ domain: demo.nebari.dev
-Currently, CI/CD can be setup for either [GitHub Actions](https://docs.github.com/en/actions) or [GitLab CI](https://docs.gitlab.com/ee/ci/).
+Nebari uses [infrastructure-as-code](https://en.wikipedia.org/wiki/Infrastructure_as_code) to declare a desired state for the deployment.  By using a git repository with CI/CD configured, teams can more quickly modify their deployment, empowering developers and data scientists to request the changes and have them approved by an administrator.
 
+When a `ci_cd` section is configured within your `nebari-config.yaml`, the `nebari render` command will create or update pipeline definition files which describe a CI/CD process to redeploy Nebari when changes are made to a specified branch.  Currently, Nebari can generate CI/CD for [GitHub Actions](https://docs.github.com/en/actions) and [GitLab CI](https://docs.gitlab.com/ee/ci/).
```
The use of `render` is not well streamlined to users as we want to reduce the level of abstraction as much as possible when deploying (render is called in the background before the stages begin). 

I would suggest rephrasing this to say that `deploy` will create the necessary files during the first time Nebari is deployed. And include a note saying that if the user wants to "update" the files as if running a dry-run, it can use the render.

kenafoster replied:
Updated

viniciusdc commented on docs/docs/explanations/advanced-configuration.md:
```diff
@@ -283,19 +283,32 @@ domain: demo.nebari.dev
+
+The CI/CD workflow that is best for you will depend on your organization, but the following approach will be appropriate for most situations.
+
+* An admin creates a GitLab or GitHub project, specifies a protected branch (typically `main`) and configures CI/CD variables (see note below) for the target cloud provider
```
this is more like a contextual comment then a suggestion, but there is the possibility of doing a repository auto-provision in case of hosting that on GitHub, but I am not sure if its valid commenting here, as we don't support the same for Gitlab so it would just lead to more confusion.

kenafoster replied:
OK I didn't even know that auto-provisioning a GitHub repo was a feature.  Since GitLab isn't supported, I agree that we shouldn't include it here, except maybe in a link to a different doc where it is mentioned.  Reviewing the code, looks like `repository_auto_provision` is the config key that would do this, but that key isn't anywhere in the 'nebari-docs' repo so I assume it isn't documented.

I've never tried it - is it reliable?  It isn't clear to me reviewing the source how the github Python client is supposed to authenticate to Github anyway - does it pick up on the credentials configured in your local `git` CLI?

All that is to say that I'd rather keep documenting the auto provisioning out of the scope of this PR and issue but I'll open up a follow-up issue for documenting it.

viniciusdc replied:
> All that is to say that I'd rather keep documenting the auto provisioning out of the scope of this PR and issue but I'll open up a follow-up issue for documenting it.

viniciusdc commented on docs/docs/explanations/advanced-configuration.md:
```diff
@@ -283,19 +283,32 @@ domain: demo.nebari.dev
+The CI/CD workflow that is best for you will depend on your organization, but the following approach will be appropriate for most situations.
+
+* An admin creates a GitLab or GitHub project, specifies a protected branch (typically `main`) and configures CI/CD variables (see note below) for the target cloud provider
+* The admin then creates a local copy of repository and creates the desired `nebari-config.yaml` including the `ci_cd` section
+* The admin runs `nebari render` for the first time, which in addition to generating IaC in the `stages` repo, will create either a `.github` folder or a `.gitlab-ci.yml` file containing the deploy steps.  *We strongly recommend the administrator review and understand the steps in these CI/CD files before continuing.*
```
This... It may not be as is. The intended "scenario" for a first deployment would require the user to follow either the `guided-init` or `init` itself, where the user will be prompted if they would like to use the CICD. 

Same comment regarding the use of `render` x `deploy.` In theory, the most suitable first workflow would be:
- init
- deploy
- attest the basic deployment works as intended
- add modifications as needed (including the CICD if not enabled in the initialization step)
- deploy (here, I could argue that advanced users could use the `render`)
- push (assuming the appropriate repo already exists and the vars were set)

viniciusdc replied:
I would, instead, remove the first two points there and start with the assumption that the GitLab or GitHub repo already exists and the variables are already set (as in your note), then expand into the valid settings of the `ci_cd` field (maybe move the example from bellow) in the `nebari-config.yaml`, culminating in the 
> Going forward, any user...

kenafoster replied:
I'm going to think on this a bit and follow-up with one more commit

kenafoster replied:
After more consideration, I think prescribing specific steps here may be more confusing than helpful.  The "easy way" would just be to have Nebari CLI create GitHub repos and CI/CD files for you in a flow where the deploys happen without much oversight/intervention.  However, more advanced production users (like us on our JATIC program) will want a more complex flow whose specific implementation will (a) be driven by specific organization needs, and (b) require a more advanced Nebari user than is the audience for these intro docs anyway.

So I rewrote this section to sound more like general principles than a step-by-step.  How does this sound now?

viniciusdc commented on docs/docs/get-started/installing-nebari.md:
````diff
@@ -45,6 +44,13 @@ There are currently two ways to install Nebari:
    ```
 
+:::note
+The version of the Nebari your `nebari-config.yaml` must match your Nebari deployer package version.  See [Upgrading Nebari][nebari-upgrade] for techniques for upgrading your Nebari CLI or safely updating your older config file to match your Nebari package version.
````
```suggestion
The version of Nebari in your `nebari-config.yaml` must match your current installed Nebari package version; otherwise, a warning will be raised when attempting to deploy. See [Upgrading Nebari][nebari-upgrade] for techniques for upgrading your Nebari CLI or safely updating your older config file to match your Nebari package version.
```

kenafoster replied:
Done

viniciusdc commented on docs/docs/get-started/installing-nebari.md:
```diff
@@ -45,6 +44,13 @@ There are currently two ways to install Nebari:
+:::
+
+:::note
+The Nebari CLI will auto-detect and then deploy any [Nebari Extensions][nebari-extension-system] that are installed in your Python environment, and extensions once deployed cannot be uninstalled our automatically destroyed with a `nebari destroy` command.  For this reason, we recommend creating a [unique environment][environment-management] for each Nebari deployment, especially when managing mul ...
```
This note should've been moved to the extension docs, along with an FAQ on removing extensions and managing their installs. Also, while I don't have an example to corroborate this statement, a stage extension should be able to be destroyed by terraform, as they do have a `destroy` abstract method in the Stage class to handle it.

kenafoster replied:
The "no uninstall" part of the note is already mentioned in the extension docs, but I think it's a critical enough point to mention again here, since if you mistakenly deploy an extension (especially if it is in the default namespace), removing everything that gets installed is a headache.

Regarding destroying the extension - I've never seen an extension fully destroyed.  The pods and such are of course gone when EKS is blown away but things like S3 buckets and IAM roles linger.   A quick look at `destroy()` in `NebariTerraformStage` seems to me that it should work without the extension needing to override the method... so maybe there's a bug in the hook into extensions when destroy is called.  But I'll remove that part from the doc since at best it seems like it would be documenting a bug.  Next time I destroy, I'll watch the output more closely to see if the extension is called.

viniciusdc commented on docs/docs/get-started/quickstart.md:
```diff
@@ -43,6 +44,10 @@ mamba install nebari
 </Tabs>
 
+:::note
+The Nebari CLI will auto-detect and then deploy any [Nebari Extensions][nebari-extension-system] that are installed in your Python environment, and extensions once deployed cannot be uninstalled our automatically destroyed with a `nebari destroy` command.  For this reason, we recommend creating a [unique environment][environment-management] for each Nebari deployment, especially when managing mul ...
```
I think this is a good place for this, I would just beware of the destroy part as stated above

kenafoster replied:
Removed destroy reference

viniciusdc commented on docs/docs/get-started/quickstart.md:
```diff
@@ -181,6 +186,40 @@ nebari init azure --project projectname \
+
+## Render (optional)
+
+You can generate the (Terraform) deployment workflow scripts with:
```
I would very briefly expand this a bit, as this is the actual step that [renders/loads the stage classes/models](https://github.com/nebari-dev/nebari/blob/develop/src/_nebari/render.py#L29-L33) into physical files, and then writes the data accordingly, based on what has been updated making a checksum comparison between whats present in the user workspace x a temp render location.

kenafoster replied:
Done

viniciusdc commented on docs/docs/how-tos/nebari-environment-management.md:
```diff
@@ -0,0 +1,26 @@
+Once installed, you can use [manage environments](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#managing-environments) in different ways.
+
+This is a simple way to set up a Python environment with a specific Nebari version and exactly one extension installed.  *When **pip** is installed with conda, the packages will not be saved in the global namespace*.  This especially critical when working with extensions, because `nebari` will detect and apply any extensions installed in your Python environment when it runs.
+
```
I will add a new line here, saying that we will create an environment called `nebari0`, which will "store" all of our packages for those not used to conda.

kenafoster replied:
I'm not sure what this is referring to.  I've never seen `nebari0` when I run  `conda env list` in any of my locals where I've been deploying Nebari.

What is nebari0 and how does it relate?

tylergraff replied:
I think "nebari0" is an example name for a proposed conda environment used to namespace the package dependencies needed to install Nebari.

viniciusdc replied:
Yeah, it could be anything, for example `nebari-dev' would be used to install nebari in a experimental fashion. I assume they added the extra `0` in the example `nebari0` because the text above discusses using multiple environments with possible different package versions.

kenafoster replied:
Oh my bad, I forgot that the example used `nebari0` as the conda name and that totally threw me off parsing the comment.  My second commit will include the language tweaks

viniciusdc commented on docs/docs/how-tos/nebari-environment-management.md:
```diff
@@ -0,0 +1,26 @@
+conda create -n nebari0
+conda activate nebari0
+conda install python==3.12 pip # or python version of your choice
+pip install nebari==2024.1.1 nebari-plugin-self-registration==0.0.9
```
``` suggestion
pip install nebari==2024.1.1 
# Plugins can also be installed 
pip install nebari-plugin-self-registration==0.0.9
```
I would split this part so that it's explicit that this is optional when running Nebari itself.

viniciusdc commented on docs/docs/how-tos/nebari-stages-directory.md:
```diff
@@ -0,0 +1,46 @@
+
+`nebari render` will validate your configuration file **AND** generate new Terraform manifests in the `stages` directories but will **NOT** deploy those changes.
+
+`nebari deploy` will validate your configuration file, generate new Terraform manifests **UNLESS** you use the `--disable-render`, **AND** deploy those changes.  If `--disable-render` is set, the Nebari CLI will instead use whatever code is present in the existing `stages` directory.  This is useful to ensure that only Terraform code changes that you have seen are present with the deploy (see Tip ...
```
There is another one that might be helpful displaying here as we are already talking about the extension below, `nebari info` shows all current "installed" stages (as they are also plugins) currently in place -- e.g that will be rendered/deployed:
<img width="1045" alt="Captura de Tela 2024-07-02 às 19 51 51" src="https://github.com/nebari-dev/nebari-docs/assets/51954708/baf3f92f-4d72-45f4-99fe-072831c390e7">

kenafoster replied:
Done

viniciusdc reviewed (commented):
Hi @kenafoster that looks awesome, thanks for going over and submitting these 🚀 
I've made some suggestions here and there, let me know your thoughts.

kenafoster commented on docs/docs/explanations/advanced-configuration.md:
```diff
@@ -283,19 +283,32 @@ domain: demo.nebari.dev
-Nebari uses [infrastructure-as-code](https://en.wikipedia.org/wiki/Infrastructure_as_code) to allow developers and users to request changes to the environment via pull requests (PRs) which then get approved by administrators.
-You may configure a CI/CD process to watch for pull-requests or commits on specific branches.
-Currently, CI/CD can be setup for either [GitHub Actions](https://docs.github.com/en/actions) or [GitLab CI](https://docs.gitlab.com/ee/ci/).
+Nebari uses [infrastructure-as-code](https://en.wikipedia.org/wiki/Infrastructure_as_code) to declare a desired state for the deployment.  By using a git repository with CI/CD configured, teams can more quickly modify their deployment, empowering developers and data scientists to request the changes and have them approved by an administrator.
```
OK - let me know if the re-wording in the revision commit sounds better

viniciusdc commented:
> > Hi @kenafoster this looks awesome. Thanks for going over and submitting these 🚀 I've made some suggestions here and there; let me know your thoughts.
> 
> Thank you! I've addressed most of them. One of the comments I need to give more thought to (and will send one more commit), and I'm not sure about the `nebari0` conda env you mentioned.

Excellent, thanks, @kenafoster. I will have another look today. Do not worry too much about that; it only showcases an example of utilization.

viniciusdc commented on docs/docs/get-started/quickstart.md:
```diff
@@ -11,6 +11,7 @@ This is a quick Nebari CLI reference. If you're new to Nebari, start at [Install
 </div>
 
 ## Install
```
@kenafoster, I know this is outside the scope of this PR, but could you also remove this installation section here? I think keeping this as a section here is redundant now, considering we have a whole page for that and we give a link to it at the beginning

kenafoster replied:
done

viniciusdc reviewed (commented):
Just a few more tweaks and this is good to go 🚀

kenafoster commented on docs/docs/explanations/advanced-configuration.md:
```diff
@@ -283,19 +283,33 @@ domain: demo.nebari.dev
-Currently, CI/CD can be setup for either [GitHub Actions](https://docs.github.com/en/actions) or [GitLab CI](https://docs.gitlab.com/ee/ci/).
+Nebari uses [infrastructure-as-code](https://en.wikipedia.org/wiki/Infrastructure_as_code) to maintain a description of the deployed infrastructure in source control.  By using a git repository with CI/CD configured, teams can more quickly modify their deployment, empowering developers and data scientists to request the changes and have them approved by an administrator.
 
+When a `ci_cd` section is configured within your `nebari-config.yaml`, the first `nebari deploy` command will create pipeline definition files which describe a CI/CD process to redeploy Nebari as changes are made to a specified branch.  (Alternatively, an administrator can use `nebari render` to generate the necessary files as if running a dry-run.)  Currently, Nebari can generate CI/CD for [GitH ...
```
done

kenafoster commented:
@viniciusdc thanks, ready for re-review.  Do you want to 'resolve conversation' on the comment chains where you're satisfied that I addressed your feedback?  (Unless it is 100% good to merge now, in which case doesn't matter).  Getting hard to track which to-dos are still open

viniciusdc reviewed (approved):
LGTM, thanks @kenafoster 🚀
